h2/src/main/org/h2/expression/Function.java

```diff
@@ -1863,79 +1862,70 @@ private static Timestamp dateadd(String part, long count, Timestamp d) {
      * </pre>
      *
      * @param part the part
-     * @param d1 the first date
-     * @param d2 the second date
+     * @param v1 the first date-time value
+     * @param v2 the second date-time value
      * @return the number of crossed boundaries
      */
-    private static long datediff(String part, Timestamp d1, Timestamp d2) {
+    private static long datediff(String part, Value v1, Value v2) {
         int field = getDatePart(part);
-        Calendar calendar = DateTimeUtils.createGregorianCalendar();
-        long t1 = d1.getTime(), t2 = d2.getTime();
-        // need to convert to UTC, otherwise we get inconsistent results with
-        // certain time zones (those that are 30 minutes off)
-        TimeZone zone = calendar.getTimeZone();
-        calendar.setTime(d1);
-        t1 += zone.getOffset(calendar.get(Calendar.ERA),
-                calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
-                calendar.get(Calendar.DAY_OF_MONTH),
-                calendar.get(Calendar.DAY_OF_WEEK),
-                calendar.get(Calendar.MILLISECOND));
-        calendar.setTime(d2);
-        t2 += zone.getOffset(calendar.get(Calendar.ERA),
-                calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
-                calendar.get(Calendar.DAY_OF_MONTH),
-                calendar.get(Calendar.DAY_OF_WEEK),
-                calendar.get(Calendar.MILLISECOND));
+        long[] a1 = DateTimeUtils.dateAndTimeFromValue(v1);
```

Comment:
dateAndTimeFromValue doesn't do any timezone adjustment for TimestampWithTimeZone

What happens if the two timestamps we are diff-ing are in different zones?

Reply:
Old code does not work with this type at all, so compatibility is not broken.

I think we should discuss somewhere what should we do if offsets are different. Choose first as a base, or do something else. I don't know. (If offsets are the same new code works correctly.)

Reply:
hmmm, the old code would at least throw an exception if someone did something that stupid.
I prefer we maintain that invariant for now :-)

Reply:
From SQL Server documentation:

> A time zone offset component of startdate or endate is not used in calculating the return value.

https://technet.microsoft.com/en-us/library/ms189794(v=sql.110).aspx

Reply:
Fair enough, please add that to the documentation for DATEDIFF

Reply:
This function is from SQL Server so probably we should also simply ignore offset for compatibitity.

I don't think that usage of this function with `TIMESTAMP WITH TIME ZONE` is stupid. Sometimes all values are from the same timezone and this type is used only to avoid problems on DST transition from summer to winter time and 1h of possible difference don't affect results with `MONTH` or `DAYS` arguments, for example, because there are no transitions on days boundaries.

Reply:
I did'n see your previous comment.

OK, I'll write something about it in documentation.